### DISCUSSION TITLE
Getting "Application error" page trying to use Prisma Optimize

### DISCUSSION THREAD

**Discussion by guillaumeklaus** [prisma/optimize-feedback#10] [Q&A]
### Question

I'm trying to use Prisma optimize to trace my queries but I keep getting "Application error" page reaching https://optimize.prisma.io.

My prisma client version is : "@prisma/client": "5.15.0-dev.8"

I'm using NestJS and initialize Prisma client like this : 

```
import { Injectable, OnModuleInit } from '@nestjs/common';

import { PrismaClient as PrismaClientUserData } from '@db-user-data/client';
import { withOptimize } from '@prisma/extension-optimize';

@Injectable()
export class PrismaClientUserDataService extends PrismaClientUserData implements OnModuleInit {
  constructor() {
    super().$extends(withOptimize());
  }

  async onModuleInit() {
    await this.$connect();
  }

  async onModuleDestroy() {
    await this.$disconnect();
  }
}
```

Is there something obviously wrong here ?

Thanks a lot !

- OS: macOS
- Database: PostgreSQL
- Node.js version: 18.16.0

**Comment by aqrln** [comment]
Hey @guillaumeklaus! We have identified and fixed the issue, could you please try again?

Thank you for reaching out 🙏

**Reply by guillaumeklaus** [reply]
Hey, thanks a lot @aqrln !

I can confirm my first issue is fixed, I'm now able to see the dashboard 🙌

I now have to understand why I do not see any trace on it. 
I tried to $extends using the code in my first post but since super return nothing it can't work.

I also tried something like this, but that doesn't seem to work either :
```
import { Injectable, OnModuleInit } from '@nestjs/common';

import { PrismaClient as PrismaClientDvf } from '@db-dvf/client';
import { withOptimize } from '@prisma/extension-optimize';

@Injectable()
export class PrismaClientDvfService extends PrismaClientDvf implements OnModuleInit {
  async onModuleInit() {
    this.$extends(withOptimize());
    await this.$connect();
  }

  async onModuleDestroy() {
    await this.$disconnect();
  }
}
```

Do you have an idea ? 🙏

**Reply by aqrln** [reply]
`this.$extends` doesn't mutate state, it returns a new, extended client.

You can look at this issue: https://github.com/prisma/prisma/issues/18628 to see some of the possible workarounds. Alternatively, you can use a [value provider](https://docs.nestjs.com/fundamentals/custom-providers#value-providers-usevalue) or a [factory provider](https://docs.nestjs.com/fundamentals/custom-providers#factory-providers-usefactory) in NestJS instead of a class provider.

**Reply by aqrln** [reply]
I'm personally a bit skeptical that it is a useful DI boundary in the first place (since you aren't going to have another implementation of `PrismaClient` with exactly the same interface) and I'd just export a singleton client from some module and use it in the services managing the specific entities directly (which, themselves, could have alternative implementations like mocks for unit tests that don't use the database and thus Prisma at all) but I can also understand that if you already have a DI container anyway, then you'd prefer it to take care of lazily instantiating a singleton object instead of doing that manually.

**Reply by Mbazu-Daniel** [reply]
@aqrln did you find a solution that worked ?

**Reply by aqrln** [reply]
@Mbazu-Daniel solution to what?

**Reply by Mbazu-Daniel** [reply]
Sorry I was supposed to tag @guillaumeklaus and not you

**Comment by millsp** [comment]
This issue should be fixed now, so I will mark it as resolved. But let me know if you need anything. Also, we have released Prisma Optimize into GA, you might want to read about our new features, changes, and docs. Head over here https://www.prisma.io/data-platform/optimize.
